controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -932,3 +923,227 @@ func (r *Reconciler) attachmentsToDelete(
+				// is changed but before CNS reconciles the volume for detaching,
+				// then user could get a stale data in this case.
+				// Changing the ClaimName should be a rare case.
+				delete(orphanedAttachmentsMap, att.Name)
+			}
+		}
+	}
+
+	// For legacy attachments that are not tracked by the VM spec,
+	// add them to the volume status with detaching suffix.
+	volumeStatuses = append(volumeStatuses,
+		getVolumeStatusWithDetachingVolumeInLegacyAttachment(
+			ctx,
+			orphanedAttachmentsMap,
+		)...,
+	)
+
+	// processBatchAttachment already added the batch attachment volumes
+	// and classic volumes to the VM status.
+	return volumeStatuses
+}
+
+// updateVMVolumeStatus clears any managed volume status then adds given
+// volume status info to vm.status.volumes, then sorts the status.
+func updateVMVolumeStatus(
+	ctx *pkgctx.VolumeContext,
+	v1, v2 []vmopv1.VirtualMachineVolumeStatus,
+) {
+
+	// Remove any managed volumes from the existing status.
+	ctx.VM.Status.Volumes = slices.DeleteFunc(ctx.VM.Status.Volumes,
+		func(e vmopv1.VirtualMachineVolumeStatus) bool {
+			return e.Type != vmopv1.VolumeTypeClassic
+		})
+
+	ctx.VM.Status.Volumes = append(ctx.VM.Status.Volumes, v1...)
+	ctx.VM.Status.Volumes = append(ctx.VM.Status.Volumes, v2...)
+
+	// Sort the volume statuses to ensure consistent ordering.
+	vmopv1.SortVirtualMachineVolumeStatuses(ctx.VM.Status.Volumes)
+}
+
+// categorizeVolumeSpecs categorizes the volume specs into two categories:
+// forBatch and forLegacy.
+//  1. forBatch are the volume specs that are not tracked by legacy CnsNodeVmAttachment
+//     or the legacy attachment's PVC name does not match the current volume's PVC name.
+//  2. forLegacy are the volume specs that are tracked by legacy CnsNodeVmAttachment
+//     and the legacy attachment's PVC name matches the current volume's PVC name.
+func categorizeVolumeSpecs(
+	ctx *pkgctx.VolumeContext,
+	legacyAttachments map[string]cnsv1alpha1.CnsNodeVmAttachment,
+) (forBatch, forLegacy []vmopv1.VirtualMachineVolume) {
+
+	// Filter volumes that have PVC source and are not already tracked
+	// by legacy CnsNodeVmAttachment
+	volumeSpecsForBatch := []vmopv1.VirtualMachineVolume{}
+	volumeSpecsForLegacy := []vmopv1.VirtualMachineVolume{}
+	for _, vol := range ctx.VM.Spec.Volumes {
+		if vol.PersistentVolumeClaim == nil {
+			continue
+		}
+		if vol.PersistentVolumeClaim.UnmanagedVolumeClaim != nil {
+			// Skip unmanaged volumes.
+			continue
+		}
+		// Check if this volume is already managed by a legacy
+		// CnsNodeVmAttachment. It is safe to rely on the
+		// attachment name for the volume since that's a contract
+		// with CSI.
+		attachmentNameForVol := pkgutil.CNSAttachmentNameForVolume(ctx.VM.Name, vol.Name)
+		if legacyAttachment, ok := legacyAttachments[attachmentNameForVol]; ok {
+
+			// Need to process the volume for legacy attachment even if the PVC
+			// name changed, since need to remove it from VM volume status.
```

Comment:
I am confused by this - why isn't this regenerating the Status each time (other than the funkiness of having preserve status for Classic disks)